[tensorexpre] Add typecast when src and dest buf types are different in PlacementAllocate

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71934

Differential Revision: [D33826700](https://our.internmc.facebook.com/intern/diff/D33826700)

huiguoo committed:
[tensorexpre] Add typecast when src and dest buf types are different in PlacementAllocate

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33826700).

ZolotukhinM reviewed (commented):
The code looks good, do you mind adding a test as well?

huiguoo committed:
Update on "[tensorexpre] Add typecast when src and dest buf types are different in PlacementAllocate"


Differential Revision: [D33826700](https://our.internmc.facebook.com/intern/diff/D33826700)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33826700).

ZolotukhinM commented on test/cpp/tensorexpr/test_memplanning.cpp:
```diff
@@ -85,6 +85,195 @@ TEST(BufLiveRange, MulRangeLine) {
+            CT.load(m, n), zero, zero, CT.load(m, n), kLT);
+      });
+
+  VarHandle i("i", kInt);
+  VarHandle j("j", kInt);
+  StorePtr store_e = Store::make(
+      EP, {i, j}, Cast::make(kQUInt8, DT.load(i, j) + DT.load(i, j)));
+  StmtPtr loop_e = For::make(i, 0, 32, For::make(j, 0, 32, store_e));
+
+  StorePtr store_f =
+      Store::make(FP, {i, j}, Load::make(EP, {i, j}) + Load::make(EP, {i, j}));
+  StmtPtr loop_f = For::make(i, 0, 32, For::make(j, 0, 32, store_f));
+  auto stmt = Block::make({CT.stmt(), DT.stmt(), loop_e, loop_f});
```
This looks weird to mix `Compute` API with constructing IR by hand. Why can't we use `Compute` for EP and FP?

huiguoo replied:
There is a `cast` in the store statement. I tried `Compute`, and it does not generate `cast`s automatically. Will have to manually construct it.

ZolotukhinM replied:
You don't need it to generate `cast` automatically - you specify the expression yourself there and can include `cast` too.

ZolotukhinM commented on test/cpp/tensorexpr/test_memplanning.cpp:
```diff
@@ -85,6 +85,195 @@ TEST(BufLiveRange, MulRangeLine) {
+  //   }
+  // }
+
+  SimpleIREvaluator cg(stmt, {AP, BP, FP});
```
Since we've already constructed a codegen object here, let's run it to make sure the produced results are valid?

huiguoo replied:
The produced results are not very related to the management of allocations/frees. I can add a check for the results in another PR though. Currently all tests in this file only check IR.

ZolotukhinM replied:
When we compare execution results, we make sure that the compiled and optimized version runs correctly (i.e. we don't check numerical accuracy, but we check that there is no seg-faults etc).

ZolotukhinM commented on test/cpp/tensorexpr/test_memplanning.cpp:
```diff
@@ -85,6 +85,195 @@ TEST(BufLiveRange, MulRangeLine) {
+#endif
+}
+
+TEST(MemPlanning, NoMemReuseForLargerType) {
```
Same remarks apply to this test.

huiguoo committed:
Update on "[tensorexpre] Add typecast when src and dest buf types are different in PlacementAllocate"


Differential Revision: [D33826700](https://our.internmc.facebook.com/intern/diff/D33826700)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33826700).

ZolotukhinM commented on test/cpp/tensorexpr/test_memplanning.cpp:
```diff
@@ -85,6 +87,253 @@ TEST(BufLiveRange, MulRangeLine) {
+  int N = 32;
+  int K = 32;
+
+  BufHandle AP("A", {M, K}, kShort);
+  BufHandle BP("B", {K, N}, kShort);
+  BufHandle CP("C", {M, N}, kShort);
+  BufHandle DP("D", {M, N}, kShort);
+  BufHandle EP("E", {M, N}, kFloat);
+  BufHandle FP("F", {M, N}, kFloat);
```
This test is almost identical to the previous one. Can we refactor them to avoid code duplication?

huiguoo replied:
The stmt construction process is reusable. I tried making it a common process for the two tests but the code is ugly because we need both the constructed stmt and the input/output buffers for checking. The constructed stmt is checked for `Alloc/PlacementAlloc/Free` stmts and the input/output buffers are needed for codegen construction which verifies the results. The pack/unpack of stmt + output buffer is ugly, and the reuse does not save that many lines. So will keep it as two separate tests.

ZolotukhinM reviewed (approved):
Looks good! I left some additional comments below.

huiguoo commented on test/cpp/tensorexpr/test_memplanning.cpp:
```diff
@@ -85,6 +87,253 @@ TEST(BufLiveRange, MulRangeLine) {
+# CHECK: Free(D);
+# CHECK: Free(C))IR");
+
+  PaddedBuffer<float> a_v(M, K, "a");
```
Oh, it's cleaner to create a multi-dimensional array using PaddedBuffer than vector<vector>.

huiguoo commented on test/cpp/tensorexpr/test_memplanning.cpp:
```diff
@@ -85,6 +87,253 @@ TEST(BufLiveRange, MulRangeLine) {
+
+  BufHandle AP("A", {M, K}, kFloat);
+  BufHandle BP("B", {K, N}, kFloat);
+  BufHandle CP("C", {M, N}, kFloat);
+  BufHandle DP("D", {M, N}, kFloat);
+  BufHandle EP("E", {M, N}, kQUInt8);
+  BufHandle FP("F", {M, N}, kQUInt8);
+
+  VarHandle i("i", kInt);
+  VarHandle j("j", kInt);
+  VarHandle k("k", kInt);
+
+  auto zero = Cast::make(CP.node()->dtype(), 0);
+  auto store_c_init = Store::make(CP, {i, j}, zero);
+  auto store_c = Store::make(
+      CP,
+      {i, j},
+      ReduceOp::make(
+          CP.load(i, j) + AP.load(i, k) * BP.load(k, j), {k}, Sum()));
+  auto loop_c = For::make(
+      i,
+      0,
+      M,
+      For::make(
+          j, 0, N, Block::make({store_c_init, For::make(k, 0, K, store_c)})));
+
+  auto store_d = Store::make(
+      DP,
+      {i, j},
+      CompareSelect::make(CP.load(i, j), zero, zero, CP.load(i, j), kLT));
+  auto loop_d = For::make(i, 0, M, For::make(j, 0, N, store_d));
+
+  auto store_e = Store::make(
+      EP, {i, j}, Cast::make(kQUInt8, DP.load(i, j) + DP.load(i, j)));
+  auto loop_e = For::make(i, 0, M, For::make(j, 0, N, store_e));
+
+  auto store_f = Store::make(FP, {i, j}, Load::make(EP, {i, j}));
+  auto loop_f = For::make(i, 0, M, For::make(j, 0, N, store_f));
+
+  auto stmt = Block::make({loop_c, loop_d, loop_e, loop_f});
```
Thanks for the pointer! Updated.

huiguoo committed:
Update on "[tensorexpre] Add typecast when src and dest buf types are different in PlacementAllocate"


Differential Revision: [D33826700](https://our.internmc.facebook.com/intern/diff/D33826700)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33826700).

huiguoo committed:
Update on "[tensorexpre] Add typecast when src and dest buf types are different in PlacementAllocate"


Differential Revision: [D33826700](https://our.internmc.facebook.com/intern/diff/D33826700)

[ghstack-poisoned]
